File: Heaps/SmallestInfiniteSet.py
from heapq import heappop, heappush


class SmallestInfiniteSet:
    def __init__(self):
        self.current_smallest = 1
        self.in_heap = set() # Guard against duplicates in the heap.
        self.added_back = [] # Min-heap to store numbers that have been added back.

    def popSmallest(self) -> int:
        # If there are any numbers in the heap, it contains the smallest number
        if self.added_back:
            smallest = heappop(self.added_back)
            self.in_heap.remove(smallest)
            return smallest
        
        # If heap is empty, use the current smallest number (infinite tail)
        # and increment it for the next call.
        smallest = self.current_smallest
        self.current_smallest += 1
        return smallest

    def addBack(self, num: int) -> None:
        # Only add back numbers that are smaller than the current smallest number 
        # and not already in the heap (preventing duplicates).
        if num < self.current_smallest and num not in self.in_heap:
            heappush(self.added_back, num)
            self.in_heap.add(num)

# Time complexity:
# - popSmallest: O(log m) where m is the number of elements in the heap (added back numbers).
# - addBack: O(log m) for the heap push operation.

File: Heaps/SmallestInfiniteSet.java
package Heaps;

import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Set;

public class SmallestInfiniteSet {
  int smallest;
  PriorityQueue<Integer> addedBack;
  Set<Integer> inAddedBack;

  public SmallestInfiniteSet() {
    smallest = 1;
    addedBack = new PriorityQueue<>();
    inAddedBack = new HashSet<>();
  }

  // Space Complexity: O(n) for the priority queue and hash set,
  // where n is the number of elements in the addedBack min heap.

  public int popSmallest() {
    if (addedBack.isEmpty()) {
      int currSmallest = smallest;
      smallest += 1;
      return currSmallest;
    } else {
      int currSmallest = addedBack.poll();
      inAddedBack.remove(currSmallest);
      return currSmallest;
    }
  }
  // Time Complexity: O(log n) for the priority queue operations,
  // where n is the number of elements in the addedBack min heap.

  public void addBack(int num) {
    if (num < smallest && !inAddedBack.contains(num)) {
      addedBack.offer(num);
      inAddedBack.add(num);
    }
  }

  // Time Complexity: O(log n) for the priority queue insertion,
  // where n is the number of elements in the addedBack min heap.
}
